File: services/knowledge-service/src/main/java/com/poliwise/knowledge/entity/Document.java
package com.poliwise.knowledge.entity;

import com.poliwise.knowledge.enums.ChunkingStrategy;
import com.poliwise.knowledge.enums.EmbeddingModel;
import com.poliwise.knowledge.enums.FileType;
import com.poliwise.knowledge.enums.ProcessingStatus;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "documents", schema = "knowledge")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Document {

    @Id
    @Column(name = "id", nullable = false)
    private UUID id;

    @Column(name = "original_filename")
    private String originalFilename;

    @Enumerated(EnumType.STRING)
    @Column(name = "file_type", columnDefinition = "knowledge.file_type")
    private FileType fileType;

    @Column(name = "file_size_bytes")
    private Long fileSizeBytes;

    @Column(name = "mime_type")
    private String mimeType;

    @Column(name = "file_key")
    private String fileKey;

    @Column(name = "bucket_name")
    private String bucketName;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", columnDefinition = "knowledge.processing_status")
    private ProcessingStatus status;

    @Column(name = "current_version")
    private Integer currentVersion;

    @Column(name = "extracted_text", columnDefinition = "text")
    private String extractedText;

    @Column(name = "page_count")
    private Integer pageCount;

    @Column(name = "word_count")
    private Integer wordCount;

    @Column(name = "language")
    private String language;

    @Column(name = "ocr_required")
    private Boolean ocrRequired;

    @Column(name = "ocr_confidence")
    private BigDecimal ocrConfidence;

    @Enumerated(EnumType.STRING)
    @Column(name = "chunking_strategy", columnDefinition = "knowledge.chunking_strategy")
    private ChunkingStrategy chunkingStrategy;

    @Column(name = "chunk_size")
    private Integer chunkSize;

    @Column(name = "chunk_overlap")
    private Integer chunkOverlap;

    @Enumerated(EnumType.STRING)
    @Column(name = "embedding_model", columnDefinition = "knowledge.embedding_model")
    private EmbeddingModel embeddingModel;

    @Column(name = "uploaded_by")
    private UUID uploadedBy;

    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    /* ===== Helper methods cho pipeline ===== */

    public boolean isReady() {
        return status == ProcessingStatus.READY;
    }

    public boolean isProcessing() {
        return status == ProcessingStatus.PARSING || status == ProcessingStatus.CHUNKING
                || status == ProcessingStatus.EMBEDDING || status == ProcessingStatus.INDEXING;
    }
}

File: services/ingestion-service/src/models/document.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, DateTime, Integer, BigInteger, Boolean, Text, Numeric
from sqlalchemy.dialects.postgresql import UUID as PG_UUID, JSONB
from src.db.session import Base


class Document(Base):
    """SQLAlchemy model for knowledge.documents table."""
    __tablename__ = "documents"
    __table_args__ = {"schema": "knowledge"}

    id = Column(PG_UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    original_filename = Column(String(500), nullable=False)
    file_type = Column(String(50), nullable=False)
    file_size_bytes = Column(BigInteger, nullable=False)
    mime_type = Column(String(255), nullable=False)
    file_key = Column(String(500), nullable=False)
    bucket_name = Column(String(255), nullable=False, default='poliwise-documents')
    status = Column(String(50), default='PENDING')
    current_version = Column(Integer, default=1)
    extracted_text = Column(Text, nullable=True)
    page_count = Column(Integer, nullable=True)
    word_count = Column(Integer, nullable=True)
    language = Column(String(10), default='en')
    domain = Column(String(50), nullable=True)
    content_quality = Column(String(50), nullable=True)
    ocr_required = Column(Boolean, default=False)
    ocr_confidence = Column(Numeric(5, 2), nullable=True)
    chunking_strategy = Column(String(50), default='parent_child')
    chunk_size = Column(Integer, nullable=True)
    chunk_overlap = Column(Integer, nullable=True)
    embedding_model = Column(String(100), nullable=True)
    uploaded_by = Column(PG_UUID(as_uuid=True), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)
    deleted_at = Column(DateTime, nullable=True)
    expires_at = Column(DateTime, nullable=True)
    trace_id = Column(String(100), nullable=True)
    metadata_ = Column("metadata", JSONB, nullable=True, default=dict)
